UI: Enable Replay Buffer in Advanced Mode

BenJuan26 opened the pull request:
Replay buffer is currently only supported in simple mode. This change adds support to advanced mode and is, for the most part, a mirror of the addition of replay buffer to simple mode.

There are some things that were left out, such as max time, since advanced mode should always have a target bitrate. As per Jim's request, the options are in a new tab and the buffer is not available with lossless or FFmpeg enabled.

Mantis-bug: https://obsproject.com/mantis/view.php?id=792

Techjar commented:
Awesome! I will build and test this. Also, you should probably revert that merge commit and do a rebase instead, to avoid the useless spam commit.

BenJuan26 commented:
Will do, I had done a squash from a local branch to avoid that and then realized I had to do a pull.

jp9000 commented:
Thank you very much -- I won't be able to do a thorough review it right away like with the rest of the PRs.  After I'm finished with what I'm currently working on, I'll take a look along with the rest of the pull requests.  Again thanks for this by the way -- I really appreciate it.  I've been a big bogged down with my own work lately and was going to do it, and you may have saved me a lot of time.

BenJuan26 commented:
Glad to contribute. Looking forward to your review.

Techjar commented:
Small issue, by default the replay buffer is "enabled" except it doesn't actually work (start button and hotkey don't appear either). I have to disable it, apply, then enable it again and it works.

BenJuan26 commented on UI/window-basic-main.cpp:
```diff
@@ -1049,6 +1049,9 @@ bool OBSBasic::InitBasicConfigDefaults()
 	config_set_default_uint  (basicConfig, "AdvOut", "Track5Bitrate", 160);
 	config_set_default_uint  (basicConfig, "AdvOut", "Track6Bitrate", 160);
 
+	config_set_default_bool  (basicConfig, "AdvOut", "RecRB", false);
```
It's weird that the checkbox would default to checked since the default value in the config is `false`.

BenJuan26 commented:
Okay I deleted my `%AppData%/Roaming/obs-studio` folder to reset the config and I'm seeing some strange issues I hadn't run into before. For example, unchecking the replay buffer checkbox should hide the fields below and it's not doing that. I'm investigating.

BenJuan26 commented:
Fixed the stream encoder issue. Still not sure about defaulting to checked.

Techjar commented on UI/window-basic-settings.cpp:
```diff
@@ -2054,6 +2090,10 @@ void OBSBasicSettings::LoadAdvancedSettings()
 			"RecRBPrefix");
 	const char *rbSuffix = config_get_string(main->Config(), "SimpleOutput",
 			"RecRBSuffix");
+	bool replayBuf = config_get_string(main->Config(), "AdvOut",
```
Here's the problem, this should be a `config_get_bool`, whatever this is doing with a string is probably nasty.

BenJuan26 replied:
That's exactly right, I just pushed the fix

BenJuan26 commented:
Fixed the checkbox default. Sorry about the mess.

BenJuan26 commented:
Hi @RytoEX, thanks for the style remarks. I will make the necessary changes.
The only question I have is whether I'm correct in making these updates by rewriting the history of the current PR commit so that there's still only one commit going in.

RytoEX commented:
@BenJuan26 Yes, you would make the updates and force push a new single commit.  At least, that's what Jim has asked me to do in the past.

BenJuan26 commented:
I've made the style changes and fixed the settings dialog issue mentioned in the Mantis-bug where it would always default to the Advanced page even when Simple was selected. Turns out the `currentIndex` of the outputs page has to be 0 for it to work properly. I also changed the default tab in Advanced view to the first tab.

RytoEX commented on UI/window-basic-settings.cpp:
```diff
@@ -3773,6 +3828,74 @@ void OBSBasicSettings::SimpleReplayBufferChanged()
+	int vbitrate = (int)obs_data_get_int(settings, "bitrate");
+	const char *rateControl = obs_data_get_string(settings, "rate_control");
+
+	bool lossless = strcmp(rateControl, "lossless") == 0 || ui->advOutRecType->currentIndex() == 1;
```
Style:  The line length limit is a strongly preferred 80 columns.  This should be broken into two lines and properly indented.

BenJuan26 replied:
This is 78 columns in my editor. My tabs are rendered using 4 spaces. Are you assuming tabs are rendered using 8 spaces?

RytoEX commented on UI/window-basic-settings.cpp:
```diff
@@ -3773,6 +3828,74 @@ void OBSBasicSettings::SimpleReplayBufferChanged()
+
+	int64_t memMB = int64_t(seconds) * int64_t(vbitrate + abitrate) *
+			1000 / 8 / 1024 / 1024;
+	if (memMB < 1) memMB = 1;
```
Style:  I don't believe Jim uses one-line `if` statements.  Break this into two lines and properly indent it.

RytoEX commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -1185,18 +1213,25 @@ inline void AdvancedOutput::SetupRecording()
-					idx++);
+					idx);
+			if (replayBuffer)
+				obs_output_set_audio_encoder(replayBuffer, aacTrack[i], idx);
```
Style: The line length limit is a strongly preferred 80 columns. This should be broken into two lines and properly indented.

BenJuan26 replied:
This is 78 columns in my editor. My tabs are rendered using 4 spaces. Are you assuming tabs are rendered using 8 spaces?

Nightfirecat replied:
From [the coding guidelines](https://github.com/jp9000/obs-studio/blob/master/CONTRIBUTING.rst#coding-guidelines):

> Tabs for indentation, spaces for alignment. Tabs are treated as 8 columns wide.

BenJuan26 replied:
Missed that, thanks.

Techjar commented:
The settings window is still defaulting to the Video section when I open it, but with General still selected, making it require 2 clicks to get to General. Not sure what's going on there.

BenJuan26 commented:
@Techjar You're right, I had fixed the video page, but hadn't made the general page the default. It's pretty annoying that Qt Creator/Designer works that way -- if you select a page/tab it's actually making it the default. Have a look and check whether the recent push fixes the issue.

jp9000 commented:
There were a few issues with this PR:

1.) When changing the recording encoder, it wouldn't update the estimate.  I'm assuming you forgot to add the signal to that widget.  I added that signal for you.
2.) When the encoder was set to "use stream encoder", it wouldn't start up.  I'm guessing you didn't quite test that case.  I fixed that for you.
3.) When the encoding mode is set to something other than CBR/VBR/ABR, it should say that it's unable to make an estimate, and should display the option to set a maximum memory limit.  I added the maximum memory limit widget and fixed the estimate.
4.) You changed the indentation of a few lines unnecessarily.  Might want to watch out for that in the future (always do a git diff on your code to see if there's anything unusual that sticks out in your diffs like that)

I also tried to make it so that the tab itself goes invisible if FFmpeg is selected, but it seems you can't make tabs go invisible.  I guess you ran in to that same thing, so I suppose we'll just have to live with it.

Other than those issues, from my testing it all seems to be working, even with multiple audio mixes.  Good work.  Merging with those changes shortly.

jp9000 commented:
Not sure why github isn't showing PR <-> commit references anymore, this was closed via 9cba5660

kaimi commented:
> As per Jim's request, the options are in a new tab and the buffer is not available with lossless or FFmpeg enabled.

What’s the reason for that? I’d like to keep using ffmpeg output and still have a replay buffer :)

jp9000 commented:
Because the replay buffer output is an encoded output -- meaning it expects obs encoders.

kaimi commented:
> Because the replay buffer output is an encoded output -- meaning it expects obs encoders.

That does make sense :)

I know this is getting waaaayyyy off topic now, but if you allow me a last question here: I’m mainly a Linux guy so I basically defaulted to using ffmpeg, is there anything I’ll have to consider if I switch to the OBS encoder? Anything that’s not possible? I guess quality won’t drop? :)
